standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HMSHandler.java

```diff
@@ -1363,18 +1363,66 @@ private void create_database_core(RawStore ms, final Database db)
   }
 
   @Override
+  @Deprecated
   public void create_database(final Database db)
       throws AlreadyExistsException, InvalidObjectException, MetaException {
-    startFunction("create_database", ": " + db.toString());
+    CreateDatabaseRequest req = new CreateDatabaseRequest();
+    req.setDatabaseName(db.getName());
+    if (db.isSetDescription()) {
+      req.setDescription(db.getDescription());
+    }
+    if (db.isSetLocationUri()) {
+      req.setLocationUri(db.getLocationUri());
+    }
+    if (db.isSetParameters()) {
+      req.setParameters(db.getParameters());
+    }
+    if (db.isSetPrivileges()) {
+      req.setPrivileges(db.getPrivileges());
+    }
+    if (db.isSetOwnerName()) {
+      req.setOwnerName(db.getOwnerName());
+    }
+    if (db.isSetOwnerType()) {
+      req.setOwnerType(db.getOwnerType());
+    }
+    if (db.isSetCatalogName()) {
+      req.setCatalogName(db.getCatalogName());
+    }
+    if (db.isSetCreateTime()) {
+      req.setCreateTime(db.getCreateTime());
+    }
+    if (db.isSetManagedLocationUri()) {
+      req.setManagedLocationUri(db.getManagedLocationUri());
+    }
+    if (db.isSetType()) {
+      req.setType(db.getType());
+    }
+    if (db.isSetConnector_name()) {
+      req.setDataConnectorName(db.getConnector_name());
+    }
+    if (db.isSetRemote_dbname()) {
+      req.setRemote_dbname(db.getRemote_dbname());
+    }
+    create_database_req(req);
```

Comment:
Maybe it better to construct the database from `CreateDatabaseRequest` then call `create_database(Database)`. From my point of view, in `create_database_req(req)`, we must retrieve the attributes again from `req`, it's kinds of duplicate, and `CreateDatabaseRequest` is meant to convey messages over wire, if we have already reached the HMS side, there is no need to create it again.
Another benefit is that we can make the new changes less intervened with the elder implements and better for review, what do you think?

Reply:
Again, `create_database(final Database db)` api is a deprecated method, we don't want to do anything in this api other than re-routing it to `create_database_req(req)` api. In the next release, we would like to remove all the deprecated APIs in the thrift class. Even though it breaks the incompatibility, moving forward we'll have a  clean code base.